### PR TITLE
Task 6: Redux

### PR DESCRIPTION
**Pull request by Olga2157**
Added Redux, React-Redux, Redux-Thunk.

Implemented searching, creating, editing and updating films.
Activated filtering and sorting (by genre, highest rating, newest release date).
Updated pagination with relevant movies.

Additional tools: camelcase-keys https://www.npmjs.com/package/camelcase-keys 

Useful links: 
GitHub Repository with local Movies API Service implementation: https://github.com/VarvaraZadnepriak/MoviesAPI.ReactJS
Movies API Service is available on http://localhost:4000/.
API Swagger documentation is available on http://localhost:4000/api-docs.

### CODE REVIEW

File: src/components/HomePage/FindSection.tsx [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
 import { Input, InputGroup } from 'reactstrap';
+import { getMoviesCreator } from '../../redux/actions/actionCreators/actionCreators';
 import { AppButton, Heading } from '../shared';
 
-export const FindSection: FC = () => (
+export const FindSection: FC = () => {
+  const dispatch = useDispatch();
+
+  const search = useCallback(() => {
+    const input = document.getElementById('input-find-section') as HTMLInputElement;
```
Comment (Djidi1) [inline src/components/HomePage/FindSection.tsx]: useRef instead

Comment (Djidi1) [reply src/components/HomePage/FindSection.tsx]: or state with a value:

const [input, setInput] = useState(''); // '' is the initial state value
...
    <input value={input} onInput={e => setInput(e.target.value)}/>

File: src/components/HomePage/Forms/MovieForm.tsx [hunk 2]
```diff
@@ -1,32 +1,77 @@
...
 import { InputWithLabel } from './InputWithLabel';
 import { SelectInputGenre } from './SelectInputGenre';
 
-export const MovieForm: FC<{ movieInfo?: MovieInfo }> = ({ movieInfo }) => {
+export const MovieForm: FC<{ movieInfo?: MovieInfoDetails }> = ({ movieInfo }) => {
+  const dispatch = useDispatch();
+
+  const submit = (e: FormEvent<HTMLElement>) => {
+    e.preventDefault();
+    // todo check if mandatory fields filled
```
Comment (Djidi1) [inline src/components/HomePage/Forms/MovieForm.tsx:22]: Let's use https://formik.org/ for form validation, and etc.

File: src/components/HomePage/Forms/MovieForm.tsx [hunk 3]
```diff
@@ -1,32 +1,77 @@
...
+    const title = (document.getElementById('movieTitle') as HTMLTextAreaElement).value;
```
Comment (Djidi1) [inline src/components/HomePage/Forms/MovieForm.tsx]: Anyway we do not need to use js selectors to get values:

const [title, setTitle] = useState('')
...
<input onChange={event => setTitle(event.target.value)} />

File: src/components/HomePage/Modals/MovieModal.tsx [hunk 4]
```diff
@@ -5,25 +5,25 @@ import { ButtonType } from '../../../model/enums/ButtonType';
...
-    }
-    toggleModal();
-  }, [onClickCallBack, toggleModal]);
+    const [isOpen, toggleModal] = useToggle(false);
+    const onClick = useCallback(() => {
+      if (onClickCallBack) {
+        onClickCallBack();
+      }
+      toggleModal();
+    }, [onClickCallBack, toggleModal]);
```
Comment (Djidi1) [inline src/components/HomePage/Modals/MovieModal.tsx]: we do not need to pass functions as dependencies

File: src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx [hunk 5]
```diff
@@ -1,37 +1,65 @@
...
 export const PaginationComponent: FC = () => {
-  const [currentPage, setPage] = useState(1);
-  const [totalPages] = useState(8);
-  // just mock data that will be updated with Redux in the next tasks
+
+  let { totalPages, totalAmount } = useSelector(moviesDataSelector);
+  const { currentPage } = useSelector(moviesDataSelector);
+  const dispatch = useDispatch();
+
+  store.subscribe(() => {
```
Comment (Djidi1) [inline src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx]: Why do we need this subscription?

File: src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx [hunk 6]
```diff
@@ -1,37 +1,65 @@
...
+      start = 1;
+      end = Math.min(totalPages + 1, 9);
+    } else if (currentPage < 8) {
+      start = 1;
+      end = Math.min(9, totalPages + 1);
+    } else {
+      start = currentPage;
+      end = Math.min(start + 8, totalPages + 1);
+    }
+    return Array.from(Array.from(Array(Math.ceil(end - start)).keys()), x => start + x);
```
Comment (Djidi1) [inline src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx]: Looks overcomplicate 
Could we simplify it?

File: src/components/MoviesSection/SortDropdown.tsx [hunk 7]
```diff
@@ -28,11 +31,17 @@ export const SortDropdown: FC = () => {
     e.stopPropagation();
     if (button.textContent) {
       setSortName(button.textContent);
+      let sortField = '';
       if (button.textContent === SortName.HIGHEST_RATING) {
-        // todo: add logic in next tasks
+        sortField = 'vote_average';
       } else if (button.textContent === SortName.NEWEST) {
-        // todo: add logic in next tasks
+        sortField = 'release_date';
```
Comment (Djidi1) [inline src/components/MoviesSection/SortDropdown.tsx]: Let's use enum for sort types

File: src/helpers/urlIsAvailable.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+export const urlIsAvailable = async (path: string): Promise<boolean> => {
```
Comment (Djidi1) [inline src/helpers/urlIsAvailable.ts]: bit unusual Promise usage...

What if?.. Something like this:

const myPromise = new Promise((resolve, reject) => {
  try {
    const response = await fetch(path)
    if (response.headers.get('content-type') === 'image/jpeg') {
      resolve(true);
    } else {
      reject(false);
    }
  } catch (err) {
    reject(err);
  }
});

Comment (Djidi1) [reply src/helpers/urlIsAvailable.ts]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

File: src/redux/actions/actionCreators/actionCreators.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+export const setCurrentPageCreator: ActionCreator<ThunkAction<
+  Promise<void>,
+  MovieInfoDetails[],
+  null,
+  UpdateCurrentPageWithMovies
+>> = (page: number) => {
+  return async (dispatch: Dispatch<MovieActions>) => {
+    const currentGenreFilter = store.getState().movieReducer.chosenGenre;
```
Comment (Djidi1) [inline src/redux/actions/actionCreators/actionCreators.ts:69]: a bit weird... Could we avoid `store.getState()` and pass it as param?

File: src/services/APIService.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                url += `&search=${search}`;
+            }
+            if (searchBy) {
+                url += `&searchBy=${searchBy}`;
+            }
+            if (filter && filter.toString() !== 'All') {
+                url += `&filter=${filter}`;
+            }
+        }
+        const response = await fetch(`${url}`);
```
Comment (Djidi1) [inline src/services/APIService.ts]: fetch(url)?

File: src/components/HomePage/FindSection.tsx (hunk 1 above)
Comment (Olga2157) [reply src/components/HomePage/FindSection.tsx]: added useRef in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/components/HomePage/Forms/MovieForm.tsx (hunk 3 above)
Comment (Olga2157) [reply src/components/HomePage/Forms/MovieForm.tsx]: fixed in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/components/HomePage/Modals/MovieModal.tsx (hunk 4 above)
Comment (Olga2157) [reply src/components/HomePage/Modals/MovieModal.tsx]: fixed in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx (hunk 6 above)
Comment (Olga2157) [reply src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx]: Simplified it in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/components/MoviesSection/SortDropdown.tsx (hunk 7 above)
Comment (Olga2157) [reply src/components/MoviesSection/SortDropdown.tsx]: Added enum in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/helpers/urlIsAvailable.ts (hunk 8 above)
Comment (Olga2157) [reply src/helpers/urlIsAvailable.ts]: Thanks, I fixed it in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/services/APIService.ts (hunk 10 above)
Comment (Olga2157) [reply src/services/APIService.ts]: yes, thanks. I updated this in commit e9860952613603eb2cf3095f28c180d207b9668c

File: src/components/HomePage/Forms/MovieForm.tsx (hunk 2 above)
Comment (Olga2157) [reply src/components/HomePage/Forms/MovieForm.tsx:22]: Ok, formik will be added in the next PR.

File: src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx (hunk 5 above)
Comment (Olga2157) [reply src/components/MoviesSection/PaginationComponents/PaginationComponent.tsx]: Deleted and fixed.

File: src/redux/actions/actionCreators/actionCreators.ts (hunk 9 above)
Comment (Olga2157) [reply src/redux/actions/actionCreators/actionCreators.ts:69]: discussed

File: src/components/HomePage/Forms/MovieForm.tsx (hunk 2 above)
Comment (Olga2157) [reply src/components/HomePage/Forms/MovieForm.tsx:22]: https://github.com/Olga2157/react-movies-app/pull/7

**Pull request merged**